### ISSUE TITLE
Consider to store several exit status codes

### ISSUE THREAD

**Issue by espenfl** [aiidateam/aiida-core#2890]
A feature request that would be nice to have in the release: when running nested workchains there is a need to be able to store (and later check) several entries in `exit_status`.  Maybe this is already possible, but is seems `ExitCode` is rather well defined.

**Comment by sphuber** [comment]
You mean that you want to set he exit statuses of one or more of the child processes on the calling work chain? I can see two different cases:
* You run sub processes sequentially
* You run multiple sub processes in parallel
In the first case, there is no real need for multiple exit statuses. Presumably as soon as a sub process fails there is no point in continuing and you can simply return the same exit code. Or you have some other exit code that reflects that (specific) sub process failing.

In the parallel case, you may indeed have multiple sub processes failing, but in this case, it does not make sense to me to want to attach all of them to the calling processes. It should simply return an exit code that signals one or multiple sub processes failed. The user can then always inspect the sub processes itself through the call links what their explicit exit codes were.

**Comment by espenfl** [comment]
Right now I am mostly thinking of sequential. Say we get an exit status at a low lying workchain, on which we want to act higher up the chain. However, during this act, something else happens, which causes you to return from this workchain with yet another exit stats and now it would be nice to have both, say at the top. Ideally we might not need the one at the bottom, until we fix the top one, but not sure when we would get there in practice.

**Comment by sphuber** [comment]
No, I think that every caller of a process, should *only* be concerned with the exit status of the process it directly called. It shouldn't even know about any other processes that that sub process might call. Imagine the following, `Relax` calls `Base` which calls `Calculation`. If the `Base` fails because the `Calculation` fails, the `Relax` should not have to investigate the exit status of the `Calculation`, because it shouldn't or couldn't be doing anything to fix it anyway. That is the contract of the `Base`.

**Comment by espenfl** [comment]
In principle I totally agree with this point of view. And from the numerics this is fine. 

However, when we start to introduce workchains that do more general analysis (say a workchain that verifies some physical principles based on some other exit codes, say from a `Calculation`) we might at some point need the codes from two (or more) levels in order to decipher and do a correction. We can see if we can come up with a actual example during the next months when more workchains are developed.

Also, of course it depends on how we want to use the exit statuses. Maybe we should not use them much in analysis at all, and try to analyse the data supplied instead. Then the exit status could be more coupled to the failures in the numerics. Guess it just be a design decision and we have no problem adapting to that.

You can close it for now if you want. I can reopen it if we see no way around it.

**Comment by sphuber** [comment]
I see your point, but as you say it would be good to have a use case. I implemented the exit codes really with the traditional usage in UNIX in mind. Each process terminates with one and one only. Let's just keep it open and use this ticket for discussion
